revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,232 @@
+// certificate chain, as well as errors related to these checks
+package ocsp
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	coreX509 "github.com/notaryproject/notation-core-go/x509"
+	"golang.org/x/crypto/ocsp"
+)
+
+// Options specifies values that are needed to check OCSP revocation
+type Options struct {
+	CertChain   []*x509.Certificate
+	SigningTime time.Time
+	HTTPClient  *http.Client
+}
+
+const (
+	pkixNoCheckOID      string = "1.3.6.1.5.5.7.48.1.5"
+	ocspMaxResponseSize int64  = 20480 //bytes
+)
+
+// CheckStatus checks OCSP based on the passed options and returns a 2D array of
+// errors. The size of these dimensions is as follows:
+//   - 1st dimension will always be equal in length to the certificate chain.
+//   - 2nd will have a length between 1 and the number of OCSPServers for the cert
+//
+// At each index, nil will be used to indicate no error. If there is an error,
+// it will be one of the errors defined in errors.go.
+//
+// (e.g. for a chain with 3 certs (leaf, intermediate, root) with both the leaf
+// and intermediate specifying 2 OCSP server, this could be a possible result):
+//
+//	[[PKIXNoCheckError, TimeoutError], [RevokedError], [nil]]
+func CheckStatus(opts Options) [][]error {
+	result := make([][]error, len(opts.CertChain))
+
+	if len(opts.CertChain) == 0 {
+		return result
+	}
+
+	// Validate cert chain structure
+	if len(opts.CertChain) == 1 {
+		if err := validateIssuer(opts.CertChain[0], opts.CertChain[0]); err != nil {
+			fillResultsWithError(result, InvalidChainError{IsInvalidRoot: true, Err: err})
+			return result
+		}
+	} else {
+		if err := coreX509.ValidateCodeSigningCertChain(opts.CertChain, &opts.SigningTime); err != nil {
+			fillResultsWithError(result, InvalidChainError{Err: err})
+			return result
+		}
+	}
+
+	// Check status for each cert in cert chain
+	var wg sync.WaitGroup
+	for i, cert := range opts.CertChain[:len(opts.CertChain)-1] {
+		wg.Add(1)
+		// Assume cert chain is accurate and next cert in chain is the issuer
+		go func(i int, cert *x509.Certificate) {
+			defer wg.Done()
+			result[i] = certCheckStatus(cert, opts.CertChain[i+1], opts)
+		}(i, cert)
+	}
+	// Last is root cert, which will never be revoked by OCSP
+	result[len(opts.CertChain)-1] = []error{nil}
+
+	wg.Wait()
+	return result
+}
+
+func fillResultsWithError(results [][]error, err error) {
+	for i := range results {
+		results[i] = []error{err}
+	}
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certCheckStatus(cert, issuer *x509.Certificate, opts Options) []error {
```

Comment:
Can this method return an array of ServerResults? Not clear why we still need to use 2D array of errors. We should be able to use the first-class structs instead of this 2D array, correct?